File: app/src/main/java/org/ametro/model/serialization/MetadataTypes.java
package org.ametro.model.serialization;

import com.fasterxml.jackson.databind.JsonNode;

import org.ametro.model.entities.MapDelay;
import org.ametro.model.entities.MapDelayType;
import org.ametro.model.entities.MapLocale;
import org.ametro.model.entities.MapMetadata;
import org.ametro.model.entities.MapStationInformation;
import org.ametro.model.entities.MapDelayTimeRange;
import org.ametro.model.entities.MapDelayWeekdayType;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class MetadataTypes {

    public static MapStationInformation[] asStationInformation(JsonNode arrayNode) {
        if (arrayNode == null || arrayNode.isNull()) {
            return new MapStationInformation[0];
        }

        List<MapStationInformation> array = new ArrayList<>();
        for (int i = 0; i < arrayNode.size(); i++) {
            JsonNode node = arrayNode.get(i);

            MapStationInformation scheme = new MapStationInformation(
                    node.get("line").asText(),
                    node.get("station").asText(),
                    node.get("image").asText(),
                    node.get("caption").asText(),
                    null);
            array.add(scheme);
        }
        return array.toArray(new MapStationInformation[array.size()]);
    }

    public static HashMap<Integer, String> asTextMap(JsonNode node){
        HashMap<Integer, String> textMap = new HashMap<>(node.size());

        Iterator<Map.Entry<String, JsonNode>> iterator = node.fields();
        while(iterator.hasNext()){
            Map.Entry<String, JsonNode> entry = iterator.next();
            textMap.put(Integer.parseInt(entry.getKey()), node.get(entry.getKey()).asText());
        }
        return textMap;
    }

    public static MapMetadata asMetadata(JsonNode node, MapLocale locale) {
        return new MapMetadata(
                locale,
                node.get("map_id").asText(),
                node.get("city_id").asInt(),
                node.get("timestamp").asInt(),
                node.get("latitude").asDouble(),
                node.get("longitude").asDouble(),
                asSchemeDictionary(node.get("schemes"), locale),
                asTransportSchemeDictionary(node.get("transports")),
                CommonTypes.asStringArray(node.get("transport_types")),
                asMapDelayArray(node.get("delays"), locale),
                CommonTypes.asStringArray(node.get("locales")),
                node.get("comments_text_id").asInt(),
                node.get("description_text_id").asInt(),
                node.get("file").asText()
        );
    }

    private static MapDelay[] asMapDelayArray(JsonNode arrayNode, MapLocale locale){
        if(arrayNode == null || arrayNode.isNull()){
            return new MapDelay[0];
        }
        MapDelay[] array = new MapDelay[arrayNode.size()];
        for (int i = 0; i < arrayNode.size(); i++) {
            array[i] = asMapDelay(arrayNode.get(i), locale);
        }
        return array;
    }

    private static MapDelay asMapDelay(JsonNode node, MapLocale locale){
        JsonNode nameNode = node.get("name_id");
        return new MapDelay(
                locale,
                nameNode!=null && !nameNode.isNull() ? nameNode.asInt() : null,
                asMapDelayType(node.get("type")),
                asWeekdayType(node.get("weekdays")),
                asTimeRangeArray(node.get("ranges"))
        );
    }

    private static MapDelayTimeRange[] asTimeRangeArray(JsonNode arrayNode) {
        if(arrayNode == null || arrayNode.isNull()){
            return new MapDelayTimeRange[0];
        }
        MapDelayTimeRange[] array = new MapDelayTimeRange[arrayNode.size()];
        for (int i = 0; i < arrayNode.size(); i++) {
            String[] rangeParts = arrayNode.get(i).asText().split("[:-]+");
            array[i] = new MapDelayTimeRange(
                    Integer.parseInt(rangeParts[0]),
                    Integer.parseInt(rangeParts[1]),
                    Integer.parseInt(rangeParts[2]),
                    Integer.parseInt(rangeParts[3])
            );
        }
        return array;
    }

    private static MapDelayWeekdayType asWeekdayType(final JsonNode node) {
        if(node == null || node.isNull()){
            return MapDelayWeekdayType.NotDefined;
        }
        final String value = node.asText();

        switch (value){
            case "monday": return MapDelayWeekdayType.Monday;
            case "tuesday": return MapDelayWeekdayType.Tuesday;
            case "wednesday": return MapDelayWeekdayType.Wednesday;
            case "thursday": return MapDelayWeekdayType.Thursday;
            case "friday": return MapDelayWeekdayType.Friday;
            case "saturday": return MapDelayWeekdayType.Saturday;
            case "sunday": return MapDelayWeekdayType.Sunday;

            case "workdays": return MapDelayWeekdayType.Workdays;
            case "weekend": return MapDelayWeekdayType.Weekend;
        }
        throw new RuntimeException("Invalid weekday value " + value);
    }

    private static MapDelayType asMapDelayType(final JsonNode node) {
        if(node == null || node.isNull()){
            return MapDelayType.NotDefined;
        }
        final String value = node.asText();
        switch (value){
            case "custom": return MapDelayType.Custom;

            case "day": return MapDelayType.Day;
            case "night": return MapDelayType.Night;
            case "evening": return MapDelayType.Evening;
            case "mourning": return MapDelayType.Mourning;
            case "rush": return MapDelayType.Rush;

            case "direct": return MapDelayType.Direct;

            case "west-north": return MapDelayType.WestNorth;
            case "west-south": return MapDelayType.WestSouth;
            case "west-east": return MapDelayType.WestEast;

            case "east-north": return MapDelayType.EastNorth;
            case "east-south": return MapDelayType.EastSouth;
            case "east-west": return MapDelayType.EastWest;

            case "north-east": return MapDelayType.NorthEast;
            case "north-west": return MapDelayType.NorthWest;
            case "north-south": return MapDelayType.NorthSouth;

            case "south-east": return MapDelayType.SouthEast;
            case "south-west": return MapDelayType.SouthWest;
            case "south-north": return MapDelayType.SouthNorth;
        }
        throw new RuntimeException("Invalid delay type value " + value);
    }


    private static Map<String, MapMetadata.Scheme> asSchemeDictionary(JsonNode arrayNode, MapLocale locale) {
        if (arrayNode == null || arrayNode.isNull()) {
            return new HashMap<>();
        }
        HashMap<String, MapMetadata.Scheme> dict = new HashMap<>();
        for (int i = 0; i < arrayNode.size(); i++) {
            JsonNode node = arrayNode.get(i);

            MapMetadata.Scheme scheme = new MapMetadata.Scheme(
                    locale,
                    node.get("name").asText(),
                    node.get("name_text_id").asInt(),
                    node.get("type_name").asText(),
                    node.get("type_text_id").asInt(),
                    node.get("file").asText(),
                    CommonTypes.asStringArray(node.get("transports")),
                    CommonTypes.asStringArray(node.get("default_transports")),
                    node.get("root").asBoolean()
            );
            dict.put(scheme.getName(), scheme);
        }
        return dict;
    }

    private static Map<String, MapMetadata.TransportScheme> asTransportSchemeDictionary(JsonNode arrayNode) {
        if (arrayNode == null || arrayNode.isNull()) {
            return new HashMap<>();
        }
        HashMap<String, MapMetadata.TransportScheme> dict = new HashMap<>();
        for (int i = 0; i < arrayNode.size(); i++) {
            JsonNode node = arrayNode.get(i);

            MapMetadata.TransportScheme scheme = new MapMetadata.TransportScheme(
                    node.get("name").asText(),
                    node.get("file").asText(),
                    node.get("type").asText()
            );
            dict.put(scheme.getName(), scheme);
        }
        return dict;
    }

}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/model/serialization/MetadataTypes.kt
package io.github.romangolovanov.apps.ametro.model.serialization

import com.fasterxml.jackson.databind.JsonNode

import io.github.romangolovanov.apps.ametro.model.entities.MapDelay
import io.github.romangolovanov.apps.ametro.model.entities.MapDelayTimeRange
import io.github.romangolovanov.apps.ametro.model.entities.MapDelayType
import io.github.romangolovanov.apps.ametro.model.entities.MapDelayWeekdayType
import io.github.romangolovanov.apps.ametro.model.entities.MapLocale
import io.github.romangolovanov.apps.ametro.model.entities.MapMetadata
import io.github.romangolovanov.apps.ametro.model.entities.MapStationInformation

object MetadataTypes {

    fun asStationInformation(arrayNode: JsonNode?): Array<MapStationInformation> {
        if (arrayNode == null || arrayNode.isNull) return emptyArray()
        return Array(arrayNode.size()) { i ->
            val node = arrayNode.get(i)
            MapStationInformation(
                node.get("line").asText(),
                node.get("station").asText(),
                node.get("image").asText(),
                node.get("caption").asText(),
                null
            )
        }
    }

    fun asTextMap(node: JsonNode): Map<Int, String> {
        return node.fields().asSequence()
            .associate { entry -> entry.key.toInt() to node.get(entry.key).asText() }
    }

    fun asMetadata(node: JsonNode, locale: MapLocale?): MapMetadata {
        return MapMetadata(
            locale,
            node.get("map_id").asText(),
            node.get("city_id").asInt(),
            node.get("timestamp").asInt(),
            node.get("latitude").asDouble(),
            node.get("longitude").asDouble(),
            asSchemeDictionary(node.get("schemes"), locale),
            asTransportSchemeDictionary(node.get("transports")),
            CommonTypes.asStringArray(node.get("transport_types")),
            asMapDelayArray(node.get("delays"), locale),
            CommonTypes.asStringArray(node.get("locales")),
            node.get("comments_text_id").asInt(),
            node.get("description_text_id").asInt(),
            node.get("file").asText()
        )
    }

    private fun asMapDelayArray(arrayNode: JsonNode?, locale: MapLocale?): Array<MapDelay> {
        if (arrayNode == null || arrayNode.isNull) return emptyArray()
        return Array(arrayNode.size()) { i -> asMapDelay(arrayNode.get(i), locale) }
    }

    private fun asMapDelay(node: JsonNode, locale: MapLocale?): MapDelay {
        val nameNode = node.get("name_id")
        return MapDelay(
            locale,
            if (nameNode != null && !nameNode.isNull) nameNode.asInt() else null,
            asMapDelayType(node.get("type")),
            asWeekdayType(node.get("weekdays")),
            asTimeRangeArray(node.get("ranges"))
        )
    }

    private fun asTimeRangeArray(arrayNode: JsonNode?): Array<MapDelayTimeRange> {
        if (arrayNode == null || arrayNode.isNull) return emptyArray()
        return Array(arrayNode.size()) { i ->
            val rangeParts = arrayNode.get(i).asText().split(Regex("[:-]+"))
            MapDelayTimeRange(
                rangeParts[0].toInt(),
                rangeParts[1].toInt(),
                rangeParts[2].toInt(),
                rangeParts[3].toInt()
            )
        }
    }

    private fun asWeekdayType(node: JsonNode?): MapDelayWeekdayType {
        if (node == null || node.isNull) return MapDelayWeekdayType.NotDefined
        return when (node.asText()) {
            "monday" -> MapDelayWeekdayType.Monday
            "tuesday" -> MapDelayWeekdayType.Tuesday
            "wednesday" -> MapDelayWeekdayType.Wednesday
            "thursday" -> MapDelayWeekdayType.Thursday
            "friday" -> MapDelayWeekdayType.Friday
            "saturday" -> MapDelayWeekdayType.Saturday
            "sunday" -> MapDelayWeekdayType.Sunday
            "workdays" -> MapDelayWeekdayType.Workdays
            "weekend" -> MapDelayWeekdayType.Weekend
            else -> throw RuntimeException("Invalid weekday value ${node.asText()}")
        }
    }

    private fun asMapDelayType(node: JsonNode?): MapDelayType {
        if (node == null || node.isNull) return MapDelayType.NotDefined
        return when (node.asText()) {
            "custom" -> MapDelayType.Custom
            "day" -> MapDelayType.Day
            "night" -> MapDelayType.Night
            "evening" -> MapDelayType.Evening
            "mourning" -> MapDelayType.Mourning
            "rush" -> MapDelayType.Rush
            "direct" -> MapDelayType.Direct
            "west-north" -> MapDelayType.WestNorth
            "west-south" -> MapDelayType.WestSouth
            "west-east" -> MapDelayType.WestEast
            "east-north" -> MapDelayType.EastNorth
            "east-south" -> MapDelayType.EastSouth
            "east-west" -> MapDelayType.EastWest
            "north-east" -> MapDelayType.NorthEast
            "north-west" -> MapDelayType.NorthWest
            "north-south" -> MapDelayType.NorthSouth
            "south-east" -> MapDelayType.SouthEast
            "south-west" -> MapDelayType.SouthWest
            "south-north" -> MapDelayType.SouthNorth
            else -> throw RuntimeException("Invalid delay type value ${node.asText()}")
        }
    }

    private fun asSchemeDictionary(arrayNode: JsonNode?, locale: MapLocale?): Map<String, MapMetadata.Scheme> {
        if (arrayNode == null || arrayNode.isNull) return emptyMap()
        return (0 until arrayNode.size()).associate { i ->
            val node = arrayNode.get(i)
            val scheme = MapMetadata.Scheme(
                locale,
                node.get("name").asText(),
                node.get("name_text_id").asInt(),
                node.get("type_name").asText(),
                node.get("type_text_id").asInt(),
                node.get("file").asText(),
                CommonTypes.asStringArray(node.get("transports")),
                CommonTypes.asStringArray(node.get("default_transports")),
                node.get("root").asBoolean()
            )
            scheme.name to scheme
        }
    }

    private fun asTransportSchemeDictionary(arrayNode: JsonNode?): Map<String, MapMetadata.TransportScheme> {
        if (arrayNode == null || arrayNode.isNull) return emptyMap()
        return (0 until arrayNode.size()).associate { i ->
            val node = arrayNode.get(i)
            val scheme = MapMetadata.TransportScheme(
                node.get("name").asText(),
                node.get("file").asText(),
                node.get("type").asText()
            )
            scheme.name to scheme
        }
    }
}
